### ISSUE TITLE
failed to select a version for the requirement `curve25519-dalek = "^2.0.0"`

### ISSUE THREAD

**Issue by EcoloSweet** [whisperfish/presage#53]
With this simple Cargo.toml
```toml
[package]
edition = "2021"
name = "project"
version = "0.1.0"

[dependencies]
presage = {git = "https://github.com/whisperfish/presage"}
```

when I run `cargo update` I have this error : 
```bash
error: failed to select a version for the requirement `curve25519-dalek = "^2.0.0"`
candidate versions found which didn't match: 3.2.1
location searched: Git repository https://github.com/signalapp/curve25519-dalek.git?branch=lizard2
required by package `zkgroup v0.7.3 (https://github.com/signalapp/zkgroup?tag=v0.7.3#197c382e)`
    ... which satisfies git dependency `zkgroup` of package `libsignal-service v0.1.0 (https://github.com/whisperfish/libsignal-service-rs?rev=efd4ea86f57520d99141bb9c1c4b38a2ac646d6a#efd4ea86)`
    ... which satisfies git dependency `libsignal-service` of package `presage v0.2.0 (https://github.com/whisperfish/presage#8fcfb651)`
    ... which satisfies git dependency `presage` of package `project v0.1.0 (/tmp/project)`
```

Thank you !

**Comment by EcoloSweet** [comment]
Maybe @rubdos knows what happened here, seems related to https://github.com/signalapp/curve25519-dalek/pull/4

**Comment by rubdos** [comment]
You need to add a patch section to your own project.  This is how Signal builds their own too.  @gferon, you might want to add that to docs.

Currently, it's probably this:

```toml
[patch.crates-io]
curve25519-dalek = { git = 'https://github.com/signalapp/curve25519-dalek', branch = '3.0.0-lizard2' }
```

When Presage's dependencies get bumped again, you'll need the `lizard2` branch, because of the PR you linked above:

```toml
[patch.crates-io]
curve25519-dalek = { git = 'https://github.com/signalapp/curve25519-dalek', branch = 'lizard2' }
```

**Comment by EcoloSweet** [comment]
I'm afraid I have the same error, even with this "3.0.0-lizard2" patch :

```bash
error: failed to select a version for the requirement `curve25519-dalek = "^2.0.0"`
candidate versions found which didn't match: 3.2.1
location searched: Git repository https://github.com/signalapp/curve25519-dalek.git?branch=lizard2
required by package `zkgroup v0.7.3 (https://github.com/signalapp/zkgroup?tag=v0.7.3#197c382e)`
    ... which satisfies git dependency `zkgroup` of package `libsignal-service v0.1.0 (https://github.com/whisperfish/libsignal-service-rs?rev=efd4ea86f57520d99141bb9c1c4b38a2ac646d6a#efd4ea86)`
    ... which satisfies git dependency `libsignal-service` of package `presage v0.2.0 (https://github.com/whisperfish/presage#8fcfb651)`
    ... which satisfies git dependency `presage` of package `project v0.1.0 (/tmp/project)`
```

**Comment by EcoloSweet** [comment]
Be I would feel more confortable if I didn't have to patch this dependency, as some other crates of my project (not related to signal) are using curve25519-dalek too and I don't want them to fail because of this patch :-/

**Comment by EcoloSweet** [comment]
I tried to clone presage, update the dependency "libsignal-service" and "libsignal-service-hyper", and add the patch (so, in presage Cargo.toml), cargo update is ok, but I tryed to run "cargo check" and have this error : 

```bash
error[E0599]: the method `is_identity` exists for struct `MontgomeryPoint`, but its trait bounds were not satisfied
   --> /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/x25519-dalek-1.2.0/src/x25519.rs:265:17
    |
265 |         !self.0.is_identity()
    |                 ^^^^^^^^^^^ method cannot be called on `MontgomeryPoint` due to unsatisfied trait bounds
    |
   ::: /home/user/.cargo/git/checkouts/curve25519-dalek-72abe9b51f92c496/b3329ed/src/montgomery.rs:70:1
    |
70  | pub struct MontgomeryPoint(pub [u8; 32]);
    | -----------------------------------------
    | |
    | doesn't satisfy `MontgomeryPoint: Identity`
    | doesn't satisfy `MontgomeryPoint: IsIdentity`
    |
    = note: the following trait bounds were not satisfied:
            `MontgomeryPoint: Identity`
            which is required by `MontgomeryPoint: IsIdentity`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `x25519-dalek` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by EcoloSweet** [comment]
If I add the "lizard2" branch patch, it throws hundreds of errors like that : 
```bash
error[E0308]: mismatched types
    --> /home/user/.cargo/git/checkouts/libsignal-client-20b116e5080385e2/916269c/rust/zkgroup/src/crypto/proofs.rs:1997:32
     |
1997 |         point_args.add("G_y1", credentials_system.G_y[1]);
     |                                ^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `curve25519_dalek::ristretto::RistrettoPoint`, found struct `RistrettoPoint`
     |
     = note: perhaps two different versions of crate `curve25519_dalek` are being used?

error[E0308]: mismatched types
    --> /home/user/.cargo/git/checkouts/libsignal-client-20b116e5080385e2/916269c/rust/zkgroup/src/crypto/proofs.rs:1998:32
     |
1998 |         point_args.add("G_y2", credentials_system.G_y[2]);
     |                                ^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `curve25519_dalek::ristretto::RistrettoPoint`, found struct `RistrettoPoint`
     |
     = note: perhaps two different versions of crate `curve25519_dalek` are being used?

error[E0308]: mismatched types
  --> /home/user/.cargo/git/checkouts/libsignal-client-20b116e5080385e2/916269c/rust/zkgroup/src/crypto/signature.rs:38:13
   |
38 |             self.signing_key,
   |             ^^^^^^^^^^^^^^^^ expected struct `curve25519_dalek::scalar::Scalar`, found struct `Scalar`
   |
   = note: perhaps two different versions of crate `curve25519_dalek` are being used?
```

**Comment by gferon** [comment]
Could you give a try to https://github.com/whisperfish/presage/pull/55 and see if it fixes your issues? Don't forget to run `cargo update` to update your lock file before you try building.

**Comment by EcoloSweet** [comment]
Thank you

It works in my project, with this patch  (an other one curve25519-dalek fork 😅)
```toml
[patch.crates-io]
"curve25519-dalek" = { git = 'https://github.com/whisperfish/curve25519-dalek', branch = 'lizard2-update' }

[patch.'https://github.com/signalapp/curve25519-dalek']
"curve25519-dalek" = { git = 'https://github.com/whisperfish/curve25519-dalek', branch = 'lizard2-update' }
```

thank you

**Comment by gferon** [comment]
This should be fixed after merging #55 (which also contains many other improvements)
